http-gateways/DNSSEC_EXPORT_DNSLINK.md

```diff
@@ -0,0 +1,313 @@
+
+
+Validation begins in the INITIAL state. Upon reading the CAR header, we extract the CIDs and payloads of each block in the CAR (stored in a list), then enter the VALIDATE\_HASH state. In VALIDATE\_HASH, we check that the client has support for the hash formats and codecs of the current block (and fail if not). We then hash the block payload and check that it matches the CID. If they do not match, validation fails. If they do match, we move to the next block in the list and reenter the VALIDATE\_HASH state. If there are no more blocks, transition to ACCEPT.
+
+
+```mermaid
+stateDiagram-v2
+    INITIAL --> VALIDATE_HASH: Parse blocks, [(CID, payload)]
+    INITIAL --> FAIL: Cannot parse CAR
+    VALIDATE_HASH --> VALIDATE_HASH: CID == hash(payload)
+    VALIDATE_HASH --> FAIL: CID != hash(payload)
+    VALIDATE_HASH --> ACCEPT
+```
+For the client implementation, we will prefer to use an available library. Both Go and JS libraries are available (i.e., go-car and js-car). If we opt for the browser extension design, the JS library is probably the best option.
+
+### Alternative Design
+For each request, the gateway would respond with an additional header of the form X-Ipfs-Name-Proof: <CID>, where <CID> is a pointer to an IPFS file containing the DNSSEC proof. A client could then retrieve this proof directly from IPFS if they want it. However, we decided against this approach for several reasons. The main reason is that it is never more advantageous for the gateway to retrieve the DNSSEC proof from IPFS than it would be to get it from a local cache. The only time this would be useful is if a gateway could use DNSSEC proofs found by other gateways. This is not feasible because the DNSSEC proof would be stored according to its hash (in IPFS), and the hash required can only be known by already collecting the DNSSEC proof—therefore it would be pointless for a gateway to request content that, by definition, it already has. An additional problem is that DNSSEC proofs have TTLs, and thus will not usually be valid for more than an hour. This limits the time that the content would be useful anyway.
+
+
+
+## Data Model
+In this section we define the data model and structures that define the serialization format. All data will be serialized in the order specified in the following structs, without padding. All multi-byte integers are encoded big-endian. The top-level structure is IpfsDnssecChain, all data necessary for a client to validate the resolution's correctness is nested within this structure.
+
+```c
+struct IpfsDnssecChain {
+  uint8 version;
+  DnssecProof proof;
+  Car ipfsData;
+}
```

Comment:
coupling has the advantage of not requiring state, nor a two requests protocol from the client side. Similar to CAR which block headers could technically be requested separately, having them all bundled in one file is useful.
* It allows the state to be consistent when generating the bundle, and when validating it.
* Caching is also simpler as there is only one DNS/IPNS/etc view that prevails: the gateway view.

Regarding the decoupling, one could imagine a DAG build on top of IpfsDnssecChain with a node for version, a node for proof, and a node for ipfsData. Data retrieval is different from the bundling. One will likely not be able to resume download of a DNSLink proof on a gateway with a different view, but once the bundle has been generated, it can be content addressed.

Reply:
> One will likely not be able to resume download of a DNSLink proof on a gateway with a different view

Why not? Client resolved DNSLink to a CID.  CID is either a single block, or a root of a bigger DAG (e.g., UnixFS directory,).
 
If that CID is a root of a DAG, download can be not only resumed, but download of different branches of the DAG can be fetched from different gateways, or even via P2P transports.